(CODEMGMT-454) Add scaffolding for R10K::API functions.

scotje opened the pull request:
The generated YARD doc is probably the easiest way to digest this, but I'm not sure if there's a better way to see that than just checking out this branch and running `bundle exec yard server`

scotje committed:
(CODEMGMT-454) Add scaffolding for R10K::API functions.

adreyer commented on lib/r10k/api.rb:
```diff
@@ -0,0 +1,160 @@
+    # @param opts [Hash] Additional options as defined.
+    # @return [true] Returns true on success, raises on failure.
+    # @raise [RuntimeError] Something bad happened!
+    def write_module!(module_slug, env_map, path, opts={})
```
module_slug concerns me since r10k currently works with short names as well.

Why does this need a full env_map?

scotje replied:
You're correct, at this level it should just be module_name.

It doesn't need to be a full env_map, but I feel like the API is easier to consume if this function handles destructuring what it needs out of the map instead of making the consuming code do that.

adreyer commented on lib/r10k/api.rb:
```diff
@@ -0,0 +1,160 @@
+
+    # Update local cache of the given remote VCS repository.
+    #
+    # @param origin [String] URI for the remote repository which should be cached or updated.
```
can we call this "remote" to line up with the config?

scotje replied:
That seems fine.

adreyer commented:
I don't like that we're referring to both control repos and module repos/forge as "source"

I think we should probably add:
resolve_module
deploy_module(or something) which links cache updating, resolution and writing. I know we won't use this now but it seems like a good check to make sure the library is usable outside of the specific update an environment workflow.

scotje committed:
Fixups from code review.

scotje commented:
I added resolve_module.

I can add some end-to-end functions but would it be more useful to maybe define the workflows/operations we want to make sure the API supports and maybe create some sort of integration testing around those?

adreyer commented:
I think deploy/update this module in this environment is probably the main one.

scotje committed:
More code review fixups and added the end-to-end implementations.

scotje commented:
Updated with source -> module_source renaming where appropriate and added deploy_environment and deploy_module functions with sample implementations.

adreyer commented:
+1 apparently I can't merge this.
